### ISSUE TITLE
[Feature] Support link debug/release for ScalaJS

### ISSUE THREAD

**Issue by KristianAN** [oyvindberg/bleep#384]
Support link debug and release from bloop. this is not exposed through BSP, so there will need to be some bespoke solution or implement support in BSP.

scala-cli has it's own solution for this.
Some files that might be of interest there:
https://github.com/VirtusLab/scala-cli/blob/63db2b1a9c951897dd2b3cd8de01e53024317a76/modules/cli/src/main/scala/scala/cli/commands/package0/Package.scala#L947

https://github.com/VirtusLab/scala-cli/blob/main/modules/options/src/main/scala/scala/build/options/ScalaJsOptions.scala

**Comment by KristianAN** [comment]
I will look into contributing this to BSP/Bloop as a first step.

**Comment by oyvindberg** [comment]
Thanks for looking into this! I really want it myself, just haven't had the need for these platforms in a while. I originally created #14 for this, so it's been in-flight for a quite a while.

Contributing this to BSP should be the preferred option. That way it could (conceptually at least) be shared with scala-cli and others. One particular thing I have noticed which seems unfortunate for us is that bloop seems to treat `fastLink`/`fullLink` as part of the project definition instead of something which can be chosen at will. Hopefully a BSP command could at least override that.

Also wanted to note that personally I care much more about getting the functionality I do about BSP itself. So if this looks too hard, I'd be ok with another approach as well.

Two other options we could explore are:

### What scala-cli does
basically what you linked to above

### Talk to bloop in a non-standard way

Bloop already has this functionality (AFAIU), but it's exposed through the bloop cli interface and not through BSP.
There may be a path where we speak to that directly, though I have not researched if that would be viable.

**Comment by KristianAN** [comment]
Also something to note:

  template-js:
    platform:
      jsEmitSourceMaps: false
      jsJsdom: false
      jsKind: esmodule 
      jsMode: release
      jsNodeVersion: 18.4.0
      jsVersion: 1.16.0
      name: js
  and building with bleep run + vite build produces this:
  ✓ 6 modules transformed.
../server/webapp/index.html                     0.45 kB │ gzip:   0.32 kB
../server/webapp/assets/index-8vwvO0Sm.css     59.59 kB │ gzip:  10.40 kB
../server/webapp/assets/index-zo6RLal9.js   1,894.00 kB │ gzip: 325.85 kB

while this sbt config:
    scalaJSLinkerConfig ~= {
      _.withModuleKind(ModuleKind.ESModule)
        .withModuleSplitStyle(ModuleSplitStyle.SmallModulesFor(List("no.hnikt.etl.app")))
    },
    
produces:
✓ 6 modules transformed.
../server/webapp/index.html                     0.45 kB │ gzip:   0.32 kB
../server/webapp/assets/index-8vwvO0Sm.css     59.59 kB │ gzip:  10.40 kB
../server/webapp/assets/index-zo6RLal9.js   1,894.00 kB │ gzip: 325.85 kB

This is a significant difference in bundle-size. The goal should obviously be that bleep produces as small bundles as sbt. This might be a limitation of bloop though?

**Comment by KristianAN** [comment]
So bloop does not support setting modulesplitstyle, however I do not see any reason why it could not be implemented in bloop. I am guessing it just isn't there because before scalajs 1.16 there was no real reason to use esmodules as closure did not support it, as of 1.16 scalajs has it's own minifier for esmodules.

**Comment by KristianAN** [comment]
I think this issue/bug https://github.com/scalacenter/bloop/issues/1304 is still alive. I think this is blocker? You can hack around it by adding  a main object with a main method, however I don't think that is an acceptable experience.

**Comment by KristianAN** [comment]
I have opened this https://github.com/scalacenter/bloop-config/issues/87. Depending on if that get merged I think we can look into using BSP. If it's not merged there is no point in using bloop for JS linking as the bundles will be larger than the scala-cli way. 

I have also talked to @tgodzik about supporting linking through bsp, and he had an idea about using the compile interface with an extra argument to link (--with-linking), I have a rough wip branch adding this to  the bloop bsp service.

**Comment by oyvindberg** [comment]
nice! Hope we can get that through :)

**Comment by dwalend** [comment]
I've got a project where not having the javascript separate from run is starting to make me itch. 

Looks like https://github.com/scalacenter/bloop-config/pull/87 was merged in June. What's state of progress?  

Anything I can help with?

**Comment by oyvindberg** [comment]
it's shipped, thanks to @KristianAN in #445 ! check out the `link` command in the newest release. sorry we haven't been able to write much about any of the new changes yet, life is quite busy :)

**Comment by dwalend** [comment]
(And I'm not seeing it because I've got the bleep version set to 0.0.3 in the bleep.yaml file.) Thanks!

**Comment by dwalend** [comment]
The link command doesn't build the same javascript code as the run command. Should it? (The run command's works; the link command's does not. I haven't dug into it to see the difference.)

**Comment by dwalend** [comment]
I set up an isolated demo project with the scala js hello world and saw that bleep link didn't produce any javascript at all. (In my larger project it is generating something - maybe from the project dependencies.) 

bleep link does compile .scala to .class . It didn't build any javascript code. Is that expected? (In that case I still need a way to produce javascript from scala without running it via `bleep run`.)

See https://github.com/dwalend/bleepjslink/tree/main

**Comment by oyvindberg** [comment]
Can you try to run bleep configure compile-server stop-all to ensure you have up-to-date bloop running?

And check in the bleep file in that repo?

**Comment by dwalend** [comment]
> Can you try to run bleep configure compile-server stop-all to ensure you have up-to-date bloop running?

```bash
11:22 $ bleep configure compile-server stop-all 
📗 bootstrapped in 13 ms
Unexpected argument: configure

Usage:
    bleep [--no-color] [--debug] [--directory <string>] [--dev] [--no-bsp-progress] [--log-as-json] build
...
Bleeping fast build! (version 0.0.12)

```
 
> And check in the bleep file in that repo?

It's in now.

**Comment by dwalend** [comment]
> > Can you try to run bleep configure compile-server stop-all to ensure you have up-to-date bloop running?

```bash
11:29 $ bleep config compile-server stop-all
📗 bootstrapped in 13 ms
📗 stopping bloop server running at socket /Users/dwalend/Library/Caches/build.bleep/socket/graalvm-java17_22.3.1
📗 bloop server was not running at socket /Users/dwalend/Library/Caches/build.bleep/socket/graalvm-community_23.0.1
📗 bloop server was not running at socket /Users/dwalend/Library/Caches/build.bleep/socket/system
📗 bloop server was not running at socket /Users/dwalend/Library/Caches/build.bleep/socket/graalvm-community_21.0.2
```
That's better. But same results:

```bash
11:30 $ bleep clean
📗 bootstrapped in 25 ms
✔ ~/projects/bleepjslink [main|✔] 
11:32 $ bleep link
📗 bootstrapped in 7 ms
📗 bloop-rifle: Starting compilation server
📙 BSP: Setting -Wunused set to privates redundantly
📙 BSP: Browser unused value of type org.scalajs.dom.Node [code => 176, location => file:///Users/dwalend/projects/bleepjslink/Browser/src/scala/net/walend/link/Main.scala:10:4 until 10:33]
📙 BSP: Browser discarded non-Unit value of type org.scalajs.dom.Node [code => 175, location => file:///Users/dwalend/projects/bleepjslink/Browser/src/scala/net/walend/link/Main.scala:11:4 until 11:38]
📗 BSP: Generated JavaScript file '/Users/dwalend/projects/bleepjslink/.bleep/builds/normal/.bloop/Browser/Browser-js'
📗 Linking succeeded
✔ ~/projects/bleepjslink [main|✔] 
11:32 $ tree /Users/dwalend/projects/bleepjslink/.bleep/builds/normal/.bloop/Browser/Browser-js
/Users/dwalend/projects/bleepjslink/.bleep/builds/normal/.bloop/Browser/Browser-js

0 directories, 0 files
```

**Comment by KristianAN** [comment]
@dwalend The link command requires at least one @JSExport to produce output when linking to produce output. This might be an oversight in bloop. I can take a look when I have time. For now you can fix this by adding an @JSExportTopLevel to your code and it should work. https://www.scala-js.org/doc/interoperability/export-to-javascript.html.

I am guessing this has to do with the link command not specifying a main class through BSP, thus the JS linker doesn't really know what to output.

**Comment by KristianAN** [comment]
@dwalend here is a project I never got started that has a fullstack setup of bleep + vite + tailwind now with JSExportTopLevel
https://github.com/KristianAN/matplanleggeren/

**Comment by dwalend** [comment]
That's helped, but there's still some missing bits. Here's the diff between what `bleep run Browser` and `bleep link` generate.

```bash
16:07 $ diff main-from-bleep-run.js .bleep/builds/normal/.bloop/Browser/Browser-js/main.js
978,980d977
< function $s_Lnet_walend_link_Main__main__AT__V(args) {
<   $m_Lnet_walend_link_Main$().main__AT__V(args);
< }
1986d1982
< $s_Lnet_walend_link_Main__main__AT__V(new ($d_T.getArrayOf().constr)([]));
```

`bleep run` makes these extra parts.

**Comment by KristianAN** [comment]
@dwalend I posted a PR on your repr repo with how you can get it running using exports
